backend/lucene/src/main/java/org/hibernate/search/backend/lucene/orchestration/impl/LuceneParallelWorkOrchestratorImpl.java

```diff
@@ -64,7 +65,8 @@ protected void doStart(ConfigurationPropertySource propertySource) {
 	}
 
 	@Override
-	protected void doSubmit(WorkExecution<?> workExecution, OperationSubmitter operationSubmitter) {
+	protected void doSubmit(WorkExecution<?> workExecution, OperationSubmitter operationSubmitter,
+			Function<WorkExecution<?>, Runnable> blockingRetryProducer) {
 		if ( !OperationSubmitter.BLOCKING.equals( operationSubmitter ) && threads.isWriteExecutorBlocking() ) {
```

Comment:
So you mean to do something similar to what I've tried in the other Lucene orchestrator? : 
```java
if ( OperationSubmitter.rejectedExecutionException().equals( operationSubmitter ) &&
		threads.isWriteExecutorBlocking() ) {
	throw log.nonblockingOperationSubmitterNotSupported();
}
else if ( OperationSubmitter.blocking().equals( operationSubmitter ) || 
		(OperationSubmitter.rejectedExecutionException().equals( operationSubmitter ) && !threads.isWriteExecutorBlocking()) ) {
	executor.submit( workExecution );
}
else {
	// otherwise, assume that it's offloading submitter and we want to offload right away:
	operationSubmitter.submitToQueue(
			ALWAYS_BLOCKED_QUEUE,
			workExecution,
			work -> submit( work, OperationSubmitter.blocking() )
	);
}
```
Just calling the `blockingRetryProducer` won't change much as the executor we are offloading to lives within the submitter... Or I've missed your point 😃

Reply:
Yes, you're right.

However, I wouldn't use a hack based on an always blocked queue... Let's get through this.

---

So, first, I just noticed there's something wrong with `BackendThreads.isWriteExecutorBlocking()`: now that we don't necessarily rely on `ThreadPoolProvider#newScheduledExecutor` to create our write executor, it does not make sense to use `ThreadPoolProvider#isScheduledExecutorBlocking` to determine whether the executor is blocking. Maybe add a `isBlocking()` method to `SimpleScheduledExecutor` instead? For `DelegatingSimpleScheduledExecutor`, you'll probably have to pass a boolean to the constructor to be able to answer.

---

Now that we got this out of the way... instead of using an always-blocked queue, we might want to introduce a new method in `OperationSubmitter`?

E.g.:

```
	public abstract <T extends Runnable> void submitToExecutor(SimpleScheduledExecutor executor, T element,
			Function<? super T, Runnable> blockingRetryProducer) throws InterruptedException;
```

`.blocking()` would just run `executor.submit(element)`, `.rejectedExcutionException()` would throw a (meaningful) exception if `executor.isSubmitBlocking()`, and the offloading submitter would just offload all the time.

---

You may run into problems with JQAssistant because `OperationSubmitter` is API, so its methods cannot refer to an SPI type such as `SimpleScheduledExecutor`. While `OperationSubmitter` is API, we don't expect its methods to be called by users, a bit like `ElasticsearchExtension`: 

https://github.com/hibernate/hibernate-search/blob/618594c74ea2cca16ce64a17210325fe886648ab/backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/ElasticsearchExtension.java#L45-L47

So, feel free to add a similar exception for `OperationSubmitter` in JQAssistant rules. See here:

https://github.com/hibernate/hibernate-search/blob/618594c74ea2cca16ce64a17210325fe886648ab/build/jqassistant/rules.xml#L384-L385

---

Side note: we could improve event further by adding relevant methods in  `SimpleScheduledExecutor`: `submitWithoutBlockingOrReject`/`submitWithoutBlockingOrReturnNull` (the names are bad, but you see what I mean). That way each `OperationSubmitter` would have a way to do what it wants in `submitToExecutor`. But since we cannot implement `submitWithoutBlockingOrReject`/`submitWithoutBlockingOrReturnNull` with a `ScheduledExecutorService`, this improvement is probably not worth our time.